nixos/modules/installer/tools/nixos-prepare-root.sh

```diff
@@ -0,0 +1,120 @@
+#! @shell@
+
+# This script's goal is to perform all "static" setup of a filesystem structure from pre-built store paths. Everything
+# in here should run in a non-root context and inside a Nix builder. It's designed primarily to be called from image-
+# building scripts and from nixos-install, but because it makes very few assumptions about the context in which it runs,
+# it could be useful in other contexts as well.
+#
+# Current behavior:
+#  - set up basic filesystem structure
+#  - make Nix store etc.
+#  - copy Nix, system, channel, and misceallaneous closures to target Nix store
+#  - register validity of all paths in the target store
+#  - set up channel and system profiles
+
+# Ensure a consistent umask.
+umask 0022
+
+set -e
+
+mountPoint="$1"
+channel="$2"
+system="$3"
+nix_closure="$4"
+shift 4
+other_closures="$@"
+
+PATH="@coreutils@/bin:@fakechroot@/bin:@nix@/bin:@perl@/bin:@utillinux@/bin:@rsync@/bin"
+
+# fakechroot does some awkward symlink target rewriting that breaks nix-env invocations for us, so turn it off
+export FAKECHROOT_NOEXPAND_SYMLINK_TARGET=1
+
+if ! test -e "$mountPoint"; then
+    echo "mount point $mountPoint doesn't exist"
+    exit 1
+fi
+
+# Mount some stuff in the target root directory.
+mkdir -m 0755 -p $mountPoint/dev $mountPoint/proc $mountPoint/sys $mountPoint/etc $mountPoint/run $mountPoint/home
+mkdir -m 01777 -p $mountPoint/tmp
+mkdir -m 0755 -p $mountPoint/tmp/root
+mkdir -m 0755 -p $mountPoint/var
+mkdir -m 0700 -p $mountPoint/root
+
+ln -s /run $mountPoint/var/run
+
+# Create the necessary Nix directories on the target device, if they
+# don't already exist.
+mkdir -m 0755 -p \
+    $mountPoint/nix/var/nix/gcroots \
+    $mountPoint/nix/var/nix/temproots \
+    $mountPoint/nix/var/nix/userpool \
+    $mountPoint/nix/var/nix/profiles \
+    $mountPoint/nix/var/nix/db \
+    $mountPoint/nix/var/log/nix/drvs
+
+mkdir -m 1775 -p $mountPoint/nix/store
+
+# TODO: move this back
+# chown @root_uid@:@nixbld_gid@ $mountPoint/nix/store
+
+# Copy our closures to the Nix store on the target mount point, unless they're already there.
+# We copy the Nix closure first because we need to invoke its tools  to register store path validity
+for i in $nix_closure $other_closures; do
+    # We support closures both in the format produced by `nix-store --export` and by `exportReferencesGraph`,
+    # mostly because there doesn't seem to be a single format that can be produced outside of a nix build and
+    # inside one. See https://github.com/NixOS/nix/issues/1242 for more discussion.
+    if [[ "$i" =~ \.closure$ ]]; then
+        echo "importing serialized closure $i to $mountPoint..."
+        fakechroot -- chroot $mountPoint nix-store --import < $i
```

Comment:
FYI, with `1.12` I think you can just set the store to be a store at a different root without `fakechroot`